110-shared-webhooks/proposal.md

```diff
@@ -0,0 +1,171 @@
+
+For instance, the following example filters on any open or close events for merge requests to the GitLab repository `git@example.com:mike/diaspora.git`:
+
+```json
+{
+  "interface_version": "1.0",
+  "icon": "mdi:github-circle",
+  "messages": ["check"],
+  "webhooks": [
+    {
+      "type": "gitlab",
+      "payload_filter": {
+        "object_kind": "merge_request",
+        "repository": {"url": "git@example.com:mike/diaspora.git"},
+        "object_attributes": {"action": "open"}
+      }
+    },
+    {
+      "type": "gitlab",
+      "payload_filter": {
+        "object_kind": "merge_request",
+        "repository": {"url": "git@example.com:mike/diaspora.git"},
+        "object_attributes": {"action": "close"}
+      }
+    }
+  ]
+}
+```
+
+
+## Webhook Types
+
+Every webhook needs to be configured with a `type` that defines the external service that will be publishing to it. This serves two purposes:
+
+1. Namespace webhooks to the type of external service, since just using the `payload_filter` may be ambiguous (if two external services use the same payload field names)
+2. Support services that require the webhook endpoint to behave in a certain way. For instance, Slack requires the webhook endpoint to respond to a [`url_verification` request](https://api.slack.com/events/url_verification) when the webhook is first configured. Some external services may also provide special methods of proving that the webhook payload originates from a valid source (e.g. providing a special header that the endpoint must validate). Custom `types` may be implemented to support this behaviour.
+
+Note that the implementation of these webhook `types` must be built in to Concourse. Different types *may* accept different configuration options in `fly set-webhook`
+
+However, many external services don't require any special behaviour. To avoid requiring handling of every external service in Concourse, the `--type` field is just an arbitrary string. If the `type` is one of the explicitly implemented types (e.g. `slack` to support the `url_verification` request flow), the webhook endpoint will adopt that behaviour - otherwise, the type will solely be used to disambiguate webhook payloads.
+
+When the type is not explicitly handled, Concourse will still need to validate the identity of the caller, but we can typically just use tokens as we do now (see [`resource.webhook_token`]).
+
+
+## Reducing `check` Frequency
+
+In order to reduce the polling pressure of resources that are backed by a webhook, we can keep track of when the last webhook check for a resource config completed.
+
+When we try to queue a check in Lidar, we check if has been checked by a webhook. If so, we could set the default check interval to something much larger (e.g. `24h`). The intent of configuring a high interval, rather than disabling the polling entirely, is to be resilient to missed events.
+
+
+# Open Questions
+
+### Can we avoid needing to implement `types` within Concourse?
```

Comment:
One idea here would be to simply register webhooks on a team- or Concourse-install level, and then have a `webhook_ref` field under `resource` to register that resource with the global webhook handler.

Pro: doesn't require waiting for prototypes
Con: not as "clean" as automatic registration with prototypes

Reply:
Yeah, that's what I was thinking originally. However, I feel the benefits of automatic registration are super valuable in large-scale deployments with many independent teams, where orchestrating a change to hundreds/thousands of pipelines may not be so easy.

That said, maybe we could support both automatic registration and manual registration via the pipeline config. That way, we can still see some benefits without needing to wait for prototype support

Reply:
BTW, even configuring the webhook manually via the pipeline config may still require implementing some `types` within Concourse (at least with how I'm envisioning it) - e.g. a slack webhook still needs special logic to respond to the webhook challenge.

Reply:
Right, obviously there are many benefits to a prototype-driven model.

Honestly, my perspective right now is clouded by a current pipeline that I'm working on where I have a templated Terraform configuration that is currently instanced 4x with a current need to scale it to 100x and plans to scale it to 400-500x and beyond within a year, with the configuration per instance plus the template stored in the same repository. I want to have a job per instance, so that individual instances can be triggered when the configuration for an individual instance changes, without triggering the rest of the instance jobs unnecessarily. This would require different resources per instance, which doesn't scale with webhooks (as I will max out the webhook count on the repository when I hit about 20 instances), so currently I have a single resource for all of the instances which triggers every instance job. So it's rather inefficient.

Reply:
Sounds like a cool use-case, and exactly the kind that this RFC is motivated by! Some clarifying questions:
* Do all of these instances have configuration within the same repository?
* Are you looking to use a single webhook on the repository for all of these instances?
* Is each resource that's responding to these webhooks just tracking the config file for the single instance?

Assuming the answer to all of these questions is yes for the time being...

I was thinking it could look something like this:

```yaml
resources:
- name: config
  type: git
  source:
    uri: git@github.com:my-org/my-repo.git
    paths: [config/abc.yml]
  webhooks:
  - name: my-repo
    filter:
      commits:
      - modified: [config/abc.yml]
```

(I imagine a filter is important here, since otherwise every commit would trigger 400-500 checks)

Reply:
Yes, yes, and no (but basically yes). Actually we were inspired by what https://github.com/concourse/governance did with keeping most of the data in YAML, so we have a layout like
```text
data/
  templates/
    instance.yaml
    role.yaml
  roles/
    chef.yaml
    programmer.yaml
    teacher.yaml
  instance/
    foo.yaml
    bar.yaml
tf/
  locals.tf
  example.tf
```
Where, for instance, `foo` (as defined inside the YAML) might be a `chef` and a `programmer` and `bar` might be a `programmer` and a `teacher`. Terraform compiles what the different roles are supposed to mean in terms of specific Terraform resources, and creating a new instance is as simple as `cp data/templates/instance.yaml data/instance/baz.yaml` and then re-generating the pipeline (probably I will eventually move it to an instanced pipeline, especially if this PR succeeds), where a specific instance of Terraform is isolated by running e.g. `terraform workspace select foo && terraform apply -var instance_id=foo`, which allows Terraform to arbitrarily scale while staying performant by not keeping too many instances in the same state.

Reply:
I took a crack at implementing the manual form of this RFC here: https://github.com/concourse/concourse/pull/7445

Only supports team-scoped webhooks, but would be easy to extend to global. Curious to hear your thoughts, and whether you feel it would suit your needs.

Reply:
@aoldershaw Nice! I needed to go back and forth a little bit to see how it would filter the webhook events with `commits.modified` (it's not inherently obvious, and would much benefit from being documented as an example) but it looks like it fits the bill. Can't wait for a release!